pulsar-client-api/src/main/java/org/apache/pulsar/client/api/schema/GenericObject.java

```diff
@@ -0,0 +1,50 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.api.schema;
+
+import org.apache.pulsar.common.schema.SchemaType;
+
+/**
+ * This is an abstraction over the logical value that is store into a Message.
+ * Pulsar decodes the payload of the Message using the Schema that is configured for the topic.
+ */
+public interface GenericObject {
+
+    /**
+     * Return the schema tyoe.
+     *
+     * @return the schema type
+     * @throws UnsupportedOperationException if this feature is not implemented
+     * @see SchemaType#BYTES when the topic has no schema information
+     * @see SchemaType#STRING
+     * @see SchemaType#AVRO
+     * @see SchemaType#PROTOBUF_NATIVE
+     * @see SchemaType#JSON
+     */
+    SchemaType getSchemaType();
+
+    /**
+     * Return the internal native representation of the Object,
+     * like a AVRO GenericRecord, a String or a byte[].
+     *
+     * @return the decoded object
+     * @throws UnsupportedOperationException if the operation is not supported
+     */
+    Object getNativeObject();
```

Comment:
Would it be useful to use generics here so that code using this API wouldn't need to use casts that require suppressing warnings? 
something like:
```
<T> T getNativeObject(Class<T> objectType);
```

Reply:
That was my initial proposal when I send the getNativeRecord patch.
In the end we decided (with @sijie ) to not add this additional parameter because anyway the user has to use "instanceof" and/or make a cast in order to use the result.

Reply:
> anyway the user has to use "instanceof" and/or make a cast in order to use the result.

How come? Isn't that the reason to have the type parameter so that there would be no need for instanceof or casting? The expected type could be detected based on the enum from `SchemaType getSchemaType()`. No?

Reply:
@lhotari please check this discussion with @sijie on the PR that introduced getNativeRecord a couple of weeks ago
https://github.com/apache/pulsar/pull/9614#discussion_r589185278

Reply:
@sijie I checked the discussion that @eolivelli [is referring to](https://github.com/apache/pulsar/pull/9614#discussion_r589185278). 

I believe it's a common solution to have API methods such as `<T> T getNativeObject(Class<T> objectType)` . 
For example, Jackson  uses a similar solution, there is [`public <T> T readValue(String content, Class<T> valueType)` method in `ObjectMapper`](https://github.com/FasterXML/jackson-databind/blob/28862205ea6d17ae371a057cea07280ccea6416e/src/main/java/com/fasterxml/jackson/databind/ObjectMapper.java#L1545). 
Isn't this also the style of Flink Java APIs? (/cc @addisonj)

WDYT?

Reply:
@lhotari we can move this discussion in a separate issue, this patch is only about introducing GenericObject.
I am renaming the method only for consistency, because now that method may return any Java object

Reply:
@eolivelli It's fine for me. I'll open a separate issue once this PR is merged.